python/lsst/ap/verify/ingestion.py

```diff
@@ -0,0 +1,480 @@
+instruments in the future.
+"""
+
+from __future__ import absolute_import, division, print_function
+
+__all__ = ["ingestDataset"]
+
+import os
+import tarfile
+from glob import glob
+import sqlite3
+
+import lsst.log
+from lsst.obs.decam import ingest
+from lsst.obs.decam import ingestCalibs
+from lsst.obs.decam.ingest import DecamParseTask
+from lsst.pipe.tasks.ingest import IngestConfig
+from lsst.pipe.tasks.ingestCalibs import IngestCalibsConfig, IngestCalibsTask
+from lsst.pipe.tasks.ingestCalibs import IngestCalibsArgumentParser
+import lsst.daf.base as dafBase
+from lsst.ap.pipe import doIngestTemplates
+
+# Names of directories to be created in specified repository
+_INGESTED_DIR = 'ingested'
+_CALIBINGESTED_DIR = 'calibingested'
+
+# Name of defects tarball residing in dataset's defects directory
+_DEFECT_TARBALL = 'defects_2014-12-05.tar.gz'
+
+
+def ingestDataset(dataset, repository):
+    """Ingest the contents of a dataset into a Butler repository.
+
+    The original data directory shall not be modified.
+
+    Parameters
+    ----------
+    dataset : `lsst.ap.verify.dataset.Dataset`
+        The dataset to be ingested.
+    repository : `str`
+        The file location of a repository to which the data will be ingested.
+        Shall be created if it does not exist. If ``repository`` does exist, it
+        must be compatible with ``dataset`` (in particular, it must support the
+        correct instrument).
+
+    Returns
+    -------
+    metadata : `lsst.daf.base.PropertySet`
+        The full metadata from any Tasks called by this function.
+    """
+    # TODO: generalize to support arbitrary URIs (DM-11482)
+    log = lsst.log.Log.getLogger('ap.verify.ingestion.ingestDataset')
+    dataset.makeCompatibleRepo(repository)
+    log.info('Input repo at %s created.', repository)
+
+    metadata = dafBase.PropertySet()
+    metadata.combine(_ingestRaws(dataset, repository))
+    metadata.combine(_ingestCalibs(dataset, repository))
+    metadata.combine(_ingestTemplates(dataset, repository))
+    log.info('Data ingested')
+    return metadata
+
+
+def _ingestRaws(dataset, workingRepo):
+    """Ingest the raw data for use by LSST.
+
+    The original data directory shall not be modified.
+
+    Parameters
+    ----------
+    dataset : `lsst.ap.verify.dataset.Dataset`
+        The dataset on which the pipeline will be run.
+    workingRepo : `str`
+        The repository in which temporary products will be created. Must be
+        compatible with ``dataset``.
+
+    Returns
+    -------
+    metadata : `lsst.daf.base.PropertySet`
+        The full metadata from any Tasks called by this function, or `None`.
+    """
+    rawRepo = _getOutputRepo(workingRepo, _INGESTED_DIR)
+    dataFiles = _getDataFiles(dataset.rawLocation)
+    return _doIngest(rawRepo, dataset.refcatsLocation, dataFiles)
+
+
+def _ingestCalibs(dataset, workingRepo):
+    """Ingest the raw calibrations for use by LSST.
```

Comment:
Calling this "raw calibrations" is a little misleading for how they are used (we are ingesting DECam MasterCals from the community pipeline, not raw flats/biases that need to be combined before just being applied to the images). This isn't just a DECam quibble; this ingestion routine does not do anything to process the flats/biases and assumes they are ready to be applied to the raw science images as-is.